File: day0/Задача Z Конфеты/sweets_sk_cheat_clever.cpp
/**
 * Author: Sergey Kopeliovich (Burunduk30@gmail.com)
 * Date: 2013.11.15
 *
 * Comment: O(d^3), try only +-100, no overflow (use double)
 *   looks at pair and gets the third value optimal possible
 */

#include <cstdio>
#include <cassert>
#include <iostream>
#include <algorithm>

using namespace std;

#define forn(i, n) for (int i = 0; i < (int)(n); i++)
#define forab(i, a, b) for (int i = (a); i <= (b); i++)

typedef long long ll;
typedef long double dbl;

dbl best;
int rx, ry, rz;

void relax( int x, int y, int z )
{
  dbl tmp;
  if ((tmp = (dbl)x * y * z) - best > 0)
    best = tmp, rx = x, ry = y, rz = z;
}

int main()
{
  #define NAME "sweets"
  assert(freopen(NAME ".in", "r", stdin));
  assert(freopen(NAME ".out", "w", stdout));

  int n, a, b, c;
  assert(scanf("%d%d%d%d", &n, &a, &b, &c) == 4);

  const int M = 100;
  int x0 = n / (3 * a);
  int y0 = n / (3 * b);
  int z0 = n / (3 * c);
  
  rx = x0, ry = y0, rz = z0;
  best = (dbl)x0 * y0 * z0;

  forab(dx, -M, M)
    forab(dy, -M, M)
      forab(dz, -M, M)
      {
        int x = x0 + dx, y = y0 + dy, z = z0 + dz;
        if (1 <= x && 1 <= y && 1 <= z)
        {
          if ((ll)a * x + (ll)b * y <= n) relax(x, y, (n - a * x - b * y) / c);
          if ((ll)c * z + (ll)b * y <= n) relax((n - c * z - b * y) / a, y, z);
          if ((ll)a * x + (ll)c * z <= n) relax(x, (n - a * x - c * z) / b, z);
        }
      }
  
  printf("%d %d %d\n", a * rx, b * ry, c * rz);
  return 0;
}

File: Материалы для проверки решений рег этап 2015/Пробный тур/Задача Z Конфеты/sweets_sk_cheat_clever.py
#
# Author: Sergey Kopeliovich (Burunduk30@gmail.com)
# Date: 2013.11.15
#
# Comment: O(d^3), try only +-30, no overflow
#   looks at pair and gets the third value optimal possible
#

import sys

NAME = "sweets"

inf = open(NAME + ".in", "r")
out = open(NAME + ".out", "w")

n, a, b, c = list(map(int, inf.readline().split()))

def relax(x, y, z, tmp_shift):
  global best, shift, rx, ry, rz
  tmp = x * y * z
  if (tmp > best or (tmp == best and tmp_shift < shift)):
    best, rx, ry, rz, shift = tmp, x, y, z, tmp_shift
  
x0 = n // (3 * a)
y0 = n // (3 * b)
z0 = n // (3 * c)
best = x0 * y0 * z0
shift = 0
rx, ry, rz = x0, y0, z0
M = 30
r = range(-M, M + 1)
for dx in r: 
  for dy in r:
    for dz in r:
      x, y, z = x0 + dx, y0 + dy, z0 + dz
      if ((1 <= x) and (1 <= y) and (1 <= z)):
        if (a * x + b * y <= n): relax(x, y, (n - a * x - b * y) // c, max(abs(dx), abs(dy)))
        if (c * z + b * y <= n): relax((n - c * z - b * y) // a, y, z, max(abs(dz), abs(dy)))
        if (a * x + c * z <= n): relax(x, (n - a * x - c * z) // b, z, max(abs(dx), abs(dz)))

print(rx * a, ry * b, rz * c, file=out)
sys.stderr.write("optimal shift: " + str(shift) + "\n")

inf.close()
out.close()
